tokio/src/io/util/take.rs

```diff
@@ -76,24 +75,29 @@ impl<R: AsyncRead> Take<R> {
 }
 
 impl<R: AsyncRead> AsyncRead for Take<R> {
-    unsafe fn prepare_uninitialized_buffer(&self, buf: &mut [MaybeUninit<u8>]) -> bool {
-        self.inner.prepare_uninitialized_buffer(buf)
-    }
-
     fn poll_read(
         self: Pin<&mut Self>,
         cx: &mut Context<'_>,
-        buf: &mut [u8],
-    ) -> Poll<Result<usize, io::Error>> {
+        buf: &mut ReadBuf<'_>,
+    ) -> Poll<Result<(), io::Error>> {
         if self.limit_ == 0 {
-            return Poll::Ready(Ok(0));
+            return Poll::Ready(Ok(()));
         }
 
         let me = self.project();
-        let max = std::cmp::min(buf.len() as u64, *me.limit_) as usize;
-        let n = ready!(me.inner.poll_read(cx, &mut buf[..max]))?;
+        let max = std::cmp::min(buf.remaining() as u64, *me.limit_) as usize;
+        // Make a ReadBuf of the unfulled section up to max
+        let mut b = unsafe { ReadBuf::uninit(&mut buf.unfilled_mut()[..max]) };
```

Comment:
Should we add a `ReadBuf::take(&mut self, n: usize) -> ReadBuf<'_>` that does this logic? It seems useful.